sdks/java/core/src/main/java/org/apache/beam/sdk/coders/CoderRegistrar.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.coders;
+
+import com.google.auto.service.AutoService;
+import java.util.Map;
+import java.util.ServiceLoader;
+
+/**
+ * {@link Coder} creators have the ability to automatically have their
+ * {@link Coder coders} registered with this SDK by creating a {@link ServiceLoader} entry
+ * and a concrete implementation of this interface.
+ *
+ * <p>It is optional but recommended to use one of the many build time tools such as
+ * {@link AutoService} to generate the necessary META-INF files automatically.
+ */
+public interface CoderRegistrar {
```

Comment:
Does it have to be an interface for technical reasons? Otherwise I would make it an abstract class for easy backwards compatibility. I expect that we will add or alter capabilities. It should also be marked `@Experimental` (but as you know, even with the annotation, we won't break things willy-nilly).

Examples:

1. Some way of adding fallback coder providers.
2. We will probably come up with some hybrid of `CoderFactory` and `CoderProvider` so that classes can be "registered" with a provider instead of a factory. So then everything would change to be defaulted versions of that.